### PR TITLE
Register - User folder creation and more testing

### PR DESCRIPTION
**Pull request by simon-dube**

### CODE REVIEW

File: carmin-server/server/resources/authenticate.py [hunk 1]
```diff
@@ -7,24 +7,25 @@
...
 from server.common.error_codes_and_messages import INVALID_USERNAME_OR_PASSWORD
 from server.database.models.user import User
-from .decorators import marshal_request, marshal_response, login_required
+from .decorators import unmarshal_request, marshal_response, get_db_session
 from server.common.util import generate_api_key
 
 
 class Authenticate(Resource):
-    @marshal_request(AuthenticationCredentialsSchema())
+    @get_db_session
```
Comment (louis-ver) [inline carmin-server/server/resources/authenticate.py:15]: I don't think this should be a decorator. Why not simply create a scoped session from within the `post`? I find that adding too many decorators could become hard to read and add too many unnecessary arguments to the function.

Decorators above resource endpoints should be reserved to endpoint specific computations -- ie. `@login_required`, `@unmarshal_request`, `@admin_only`

File: carmin-server/server/resources/executions.py [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
 class Executions(Resource):
     def get(self):
         pass
 
-    def post(self):
-        pass
+    @login_required
+    @unmarshal_request(ExecutionSchema())
+    @marshal_response(ExecutionSchema())
+    def post(self, model, user):
```
Comment (louis-ver) [inline carmin-server/server/resources/executions.py:19]: This should probably be part of another PR. It is not completed and should not be merged into `develop`

File: carmin-server/server/resources/helpers/register.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from server.common.error_codes_and_messages import UNAUTHORIZED, PATH_EXISTS, USERNAME_ALREADY_EXISTS
+
+
+def create_user_directory(user: User) -> (Path, ErrorCodeAndMessage):
+    user_dir_absolute_path = get_user_data_directory(user)
+
+    if not is_safe_path(user_dir_absolute_path):
+        return None, UNAUTHORIZED
+
+    if os.path.exists(user_dir_absolute_path):
```
Comment (louis-ver) [inline carmin-server/server/resources/helpers/register.py]: This check is already being done in `create_directory`, no need to repeat it here

File: carmin-server/server/resources/helpers/register.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        return None, PATH_EXISTS
+
+    return create_directory(user_dir_absolute_path)
+
+
+def already_existing_username_error_code_and_message(
```
Comment (louis-ver) [inline carmin-server/server/resources/helpers/register.py]: We should write a more generic function that does this. Something that takes an `ErrorCodeAndMessage`, an `int`, and a detailed message. We do this a lot and it is quite verbose.

File: carmin-server/server/resources/models/error_code_and_message.py [hunk 5]
```diff
@@ -24,6 +24,8 @@ def __eq__(self, other):
 
 
 class ErrorCodeAndMessageSchema(Schema):
+    SKIP_VALUES = list([None])
```
Comment (louis-ver) [inline carmin-server/server/resources/models/error_code_and_message.py:27]: Do we also want to remove the empty `errorDetail` dict? I think so.

File: carmin-server/server/resources/register.py [hunk 6]
```diff
@@ -1,26 +1,40 @@
...
+from .decorators import admin_only, unmarshal_request, marshal_response, get_db_session
+from server.resources.helpers.register import create_user_directory, already_existing_username_error_code_and_message
+from server.resources.models.error_code_and_message import ErrorCodeAndMessage
+from server.common.error_codes_and_messages import UNEXPECTED_ERROR
 
 
 class Register(Resource):
     @admin_only
-    @marshal_request(AuthenticationCredentialsSchema())
+    @get_db_session
```
Comment (louis-ver) [inline carmin-server/server/resources/register.py:16]: Again, I don't think this should be a decorator

File: carmin-server/server/resources/decorators/__init__.py [hunk 7]
```diff
@@ -83,14 +96,15 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
-def admin_only(func):
+@get_db_session
```
Comment (louis-ver) [inline carmin-server/server/resources/decorators/__init__.py:99]: Should not be a decorator

File: carmin-server/server/resources/decorators/__init__.py [hunk 8]
```diff
@@ -65,15 +69,24 @@ def wrapper(*args, **kwargs):
...
-def login_required(func):
+def get_db_session(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        return func(db_session=db.create_scoped_session(), *args, **kwargs)
+
+    return wrapper
+
+
+@get_db_session
```
Comment (louis-ver) [inline carmin-server/server/resources/decorators/__init__.py:80]: Should not be a decorator

**Review by louis-ver — CHANGES_REQUESTED** [review]
Great work! Only some slight changes to make, and a few questions.

**Comment by louis-ver** [comment]
Could this branch also be rebased or merged against `develop`? I know it doesn't affect the outcome, but it would make for a cleaner PR with less commit history.

File: carmin-server/server/resources/authenticate.py (hunk 1 above)
Comment (simon-dube) [reply carmin-server/server/resources/authenticate.py:15]: I actually believe that this should be a decorator. It is a simple way for our endpoints that do require a safe database session to get one. Plus, when you look at the function, you can, in a glance, see that the function requires: an unmarshalled model, interracts with the database and is reserved to admins only.

Comment (simon-dube) [reply carmin-server/server/resources/authenticate.py:15]: Discussed and agreed to keep it for the moment. In case we feel like needing extra decorators in the future, we will reconsider its importance further and possibly move it to a utils fonction.

File: carmin-server/server/resources/register.py (hunk 6 above)
Comment (simon-dube) [reply carmin-server/server/resources/register.py:16]: Discussed and agreed to keep it for the moment. In case we feel like needing extra decorators in the future, we will reconsider its importance further and possibly move it to a utils fonction.

File: carmin-server/server/resources/decorators/__init__.py (hunk 7 above)
Comment (simon-dube) [reply carmin-server/server/resources/decorators/__init__.py:99]: Discussed and agreed to keep it for the moment. In case we feel like needing extra decorators in the future, we will reconsider its importance further and possibly move it to a utils fonction.

File: carmin-server/server/resources/decorators/__init__.py (hunk 8 above)
Comment (simon-dube) [reply carmin-server/server/resources/decorators/__init__.py:80]: Discussed and agreed to keep it for the moment. In case we feel like needing extra decorators in the future, we will reconsider its importance further and possibly move it to a utils fonction.

File: carmin-server/server/resources/executions.py (hunk 2 above)
Comment (simon-dube) [reply carmin-server/server/resources/executions.py:19]: Discussed and agreed to keep it here as it is not harmful.

File: carmin-server/server/resources/models/error_code_and_message.py (hunk 5 above)
Comment (simon-dube) [reply carmin-server/server/resources/models/error_code_and_message.py:27]: Fixed with latest push.

**Comment by louis-ver** [comment]
Excellent work.

**Review by louis-ver — APPROVED** [review]
